pallene/checker.lua

```diff
@@ -349,14 +348,60 @@ function Checker:check_program(prog_ast)
                 end
             end
 
+        elseif group_kind == "Stat" then
+            local tot_rets = 0
+            for _, tl_node in ipairs(tl_group) do
+                local stat = tl_node.stat
+                -- The toplevel return statement will be checked further on
+                if stat._tag == "ast.Stat.Return" then
+                    tot_rets = tot_rets + 1
+                    if tot_rets > 1 then
+                        type_error(stat.loc, "There can only be one toplevel return statement")
+                    end
+                else
+                    self:check_stat(stat, true)
+                end
+            end
+
         else
             error("impossible")
         end
     end
 
+    local total_nodes = #prog_ast.tls
+    local last_toplevel_node = prog_ast.tls[total_nodes]
+    if last_toplevel_node._tag ~= "ast.Toplevel.Stat" then
+        type_error(last_toplevel_node.loc, "Last Toplevel element must be a statement")
+    end
+    local last_stat = last_toplevel_node.stat
+    if last_stat._tag ~= "ast.Stat.Return" then
+        type_error(last_stat.loc, "Last Toplevel statement must be a return statement")
```

Comment:
Maybe we could merge these two error messages? In both cases we could say that the error is that the last thing in the toplevel must be a return statement.

Reply:
Agreed.